### PR TITLE
Add script to fix sql constraints by removing duplicates

### PR DESCRIPTION
**Pull request by ghost**
This script allows to fix any sql constraint for a given model.

What it does:
for the model given in parameter
1 - Find every field relating the model through a m2m or m2o relation
2 - Find duplicates
3 - For each duplicate:
        For each table containing fields defined in 1:
              Replace duplicates with the value kept

@bwrsandman

### CODE REVIEW

**Comment by ghost** [comment]
@bwrsandman Added unittests

File: openerp/openupgrade/openupgrade.py [hunk 1]
```diff
@@ -622,3 +625,167 @@ def wrapped_function(cr, version):
...
+    When upgrading the module, if there are duplicates, integrity errors
+    will be raised before the script is run but this will not prevent
+    the script from running.
+
+    :param model: the model on witch the constraint is applied
+    :param constraint_attrs: a list of string containing the fields that
+                                form the uniq key
+    """
+    pool = pooler.get_pool(cr.dbname)
+    model_table = model.replace('.', '_')
```
Comment (hbrunn) [inline openerp/openupgrade/openupgrade.py]: don't make assumptions about the table name, use `model_pool._table`

File: openerp/openupgrade/openupgrade.py [hunk 2]
```diff
@@ -622,3 +625,167 @@ def wrapped_function(cr, version):
...
+    field_pool = pool['ir.model.fields']
+    field_m2o_ids = field_pool.search(cr, SUPERUSER_ID, [
+        ('relation', '=', model),
+        ('ttype', '=', 'many2one'),
+    ])
+    # List of tables where to look for duplicates
+    # This is trivial for many2one relations
+    tables_to_lookup = [
+        (
+            field.model_id.model.replace('.', '_'),
```
Comment (hbrunn) [inline openerp/openupgrade/openupgrade.py]: see above

File: openerp/openupgrade/openupgrade.py [hunk 3]
```diff
@@ -622,3 +625,167 @@ def wrapped_function(cr, version):
...
+    # Get all fields with the given model as many2many relation
+    field_m2m_ids = field_pool.search(cr, SUPERUSER_ID, [
+        ('relation', '=', model),
+        ('ttype', '=', 'many2many'),
+    ])
+    fields_m2m = field_pool.browse(cr, SUPERUSER_ID, field_m2m_ids)
+
+    for field in fields_m2m:
+
+        other_model_table = field.model_id.model.replace('.', '_')
```
Comment (hbrunn) [inline openerp/openupgrade/openupgrade.py]: again

**Comment by ghost** [comment]
@hbrunn Fixed the table name issue. Thanks for the hint.

**Comment by bwrsandman** [comment]
:+1:

File: openerp/openupgrade/openupgrade.py [hunk 4]
```diff
@@ -22,9 +22,11 @@
 import os
 import inspect
 import logging
-from openerp import release, osv, tools, SUPERUSER_ID
+from openerp import release, osv, tools, SUPERUSER_ID, pooler
```
Comment (StefanRijnhart) [inline openerp/openupgrade/openupgrade.py]: Pooler is deprecated. Instead, we import RegistryManager in the line below.

File: openerp/openupgrade/openupgrade.py [hunk 5]
```diff
@@ -622,3 +625,167 @@ def wrapped_function(cr, version):
...
+
+
+def remove_sql_constraint_duplicates(cr, model, constraint_attrs):
+    """
+    Remove all duplicates after a sql constraint is applied on a model.
+
+    For every field many2one and many2many with the given model as relation,
+    change the duplicate ids with the id of the record kept.
+
+    This script must be called in post-migration so that the model being
```
Comment (StefanRijnhart) [inline openerp/openupgrade/openupgrade.py]: s/This script/This method/

File: openerp/addons/test_sql_constraint/README.rst [hunk 6]
```diff
@@ -0,0 +1,23 @@
+Test SQL Constraint Duplicate Remover
+=====================================
+
+This module tests the OpenUpgrade function remove_sql_constraint_duplicates
+
+The function allows to eliminate every duplicate record of any model
+given the list of attributes that form the unique sql constraint.
+
+Every records in the database of any models that refer to a removed duplicate
```
Comment (StefanRijnhart) [inline openerp/addons/test_sql_constraint/README.rst]: Typos, should be 'Every record' ... 'of any model that refers'

File: openerp/openupgrade/openupgrade.py [hunk 7]
```diff
@@ -622,3 +625,167 @@ def wrapped_function(cr, version):
...
+    edited can be accessed through the orm.
+
+    When upgrading the module, if there are duplicates, integrity errors
+    will be raised before the script is run but this will not prevent
+    the script from running.
+
+    :param model: the model on witch the constraint is applied
+    :param constraint_attrs: a list of string containing the fields that
```
Comment (StefanRijnhart) [inline openerp/openupgrade/openupgrade.py]: list of string_s_

File: openerp/openupgrade/openupgrade.py [hunk 8]
```diff
@@ -622,3 +625,167 @@ def wrapped_function(cr, version):
...
+    change the duplicate ids with the id of the record kept.
+
+    This script must be called in post-migration so that the model being
+    edited can be accessed through the orm.
+
+    When upgrading the module, if there are duplicates, integrity errors
+    will be raised before the script is run but this will not prevent
+    the script from running.
+
+    :param model: the model on witch the constraint is applied
```
Comment (StefanRijnhart) [inline openerp/openupgrade/openupgrade.py]: on which

File: openerp/openupgrade/openupgrade.py [hunk 9]
```diff
@@ -622,3 +625,167 @@ def wrapped_function(cr, version):
...
+    For every field many2one and many2many with the given model as relation,
+    change the duplicate ids with the id of the record kept.
+
+    This script must be called in post-migration so that the model being
+    edited can be accessed through the orm.
+
+    When upgrading the module, if there are duplicates, integrity errors
+    will be raised before the script is run but this will not prevent
+    the script from running.
+
```
Comment (StefanRijnhart) [inline openerp/openupgrade/openupgrade.py:643]: Can you warn about the information loss that might occur? As in 'Use with care, ...'

File: openerp/openupgrade/openupgrade.py [hunk 10]
```diff
@@ -622,3 +625,167 @@ def wrapped_function(cr, version):
...
+                    {
+                        'table_name': table_name,
+                        'field_name': table[1],
+                        'new_value': new_record_id,
+                        'old_record_ids': len(old_record_ids) > 1 and
+                        'in %s' % (tuple(old_record_ids),) or '= %s' %
+                        old_record_ids[0]
+                    })
+
+            model_pool.unlink(cr, SUPERUSER_ID, old_record_ids)
```
Comment (StefanRijnhart) [inline openerp/openupgrade/openupgrade.py:796]: Please log info about the unlinked records to track any information loss that might occur.

**Comment by StefanRijnhart** [comment]
Thanks for your work, expecially the added tests (although Travis does not run the tests on this project yet)

**Comment by ghost** [comment]
@StefanRijnhart Thanks for the review. I fixed every issue.

**Comment by StefanRijnhart** [comment]
Thanks! :+1:

**Comment by StefanRijnhart** [comment]
Merging this with two approvals and @hbrunn's complaints having been fixed.

**Pull request merged**